### PR TITLE
Don't persist downloads to the cache until they complete

### LINKED ISSUE #38: No integrity checks are ran so the broken image can be cached
If you start the download and abort it the incomplete image is saved in cache and will be returned for the consequent downloads.

This seems to be an issue since day 1 of this module.

We need the checksums for the OS images (which is doable as this module only downloads the non-configured images) and skip cache if the saved file doesn't match.

### PR DESCRIPTION
**Pull request by pimterry**
This is aiming to solve https://github.com/resin-io/resin-cli/issues/541.

It doesn't completely solve #38, since we still can't easily check the integrity of our downloads, but it does stop us permanently caching downloads that don't really complete at all.

### CODE REVIEW

File: lib/cache.coffee [hunk 1]
```diff
@@ -120,8 +120,15 @@ exports.getImageWritableStream = (deviceType, version) ->
 			# Ensure the cache directory exists, to prevent
 			# ENOENT errors when trying to write to it.
 			mkdirp(path.dirname(imagePath)).then ->
-				return fs.createWriteStream(imagePath)
+				# Append .inprogress to streams, move them to the right location only on success
+				inProgressPath = imagePath + '.inprogress'
```
Comment (jviotti) [inline lib/cache.coffee]: Would it be good to delete this file if the stream emits an `error` event?

File: lib/manager.coffee [hunk 2]
```diff
@@ -36,6 +36,7 @@ doDownload = (deviceType, version) ->
 		cache.getImageWritableStream(deviceType, version)
 		.then (cacheStream) ->
 			pass.pipe(cacheStream)
+			pass.on('end', cacheStream.persistCache)
```
Comment (jviotti) [inline lib/manager.coffee]: Should we wait until the `cacheStream.persistCache` promise is resolved before resolving this promise?

Comment (pimterry) [reply lib/manager.coffee]: That would be tidy, but we can't I think.

The promise resolves with the stream itself, and needs to resolve as soon as that starts, so we can start streaming it onwards, and show progress while the download runs. If we wait until the stream ends and the cache is persisted then we don't get that streaming goodness.

File: lib/cache.coffee (hunk 1 above)
Comment (pimterry) [reply lib/cache.coffee]: Great idea! Done, let me know what you think

**Pull request merged**